### PR TITLE
Fix Maximum call stack size exceeded in findSchemaDefinition

### LINKED ISSUE #4116: findSchemaDefinition doesn't detect cyclic references
### Prerequisites

- [X] I have searched the existing issues
- [X] I understand that providing a [SSCCE](http://sscce.org/) example is tremendously useful to the maintainers.
- [X] I have read the [documentation](https://rjsf-team.github.io/react-jsonschema-form/docs)
- [ ] Ideally, I'm providing a [sample JSFiddle](https://jsfiddle.net), [Codesandbox.io](https://codesandbox.io) or preferably a [shared playground link](https://rjsf-team.github.io/react-jsonschema-form/) demonstrating the issue.

### What theme are you using?

core

### Version

5.17.1

### Current Behavior

Consider a schema that (accidentally!) has a self-referencing definition like so:

```json
{
  "schema": {
    "definitions": {
      "oops": {
        "$ref": "#/definitions/oops",
      },
    },
  },
}
```

During validation, `findSchemaDefinition` enters a recursive loop until the runtime aborts with `Maximum call stack size exceeded` (Chrome) or `function nested too deeply` (Firefox).

### Expected Behavior

`findSchemaDefinition` could detect that the "resolved" $ref is in fact the same $ref and throw, alerting developers to the issue in their schema.

### Steps To Reproduce

_No response_

### Environment

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by MarttiR**
### Reasons for making this change

Fixes #4116 

### Checklist

- [ ] **I'm updating documentation**
  - [ ] I've [checked the rendering](https://rjsf-team.github.io/react-jsonschema-form/docs/contributing) of the Markdown text I've added
- [x] **I'm adding or updating code**
  - [x] I've added and/or updated tests. I've run `npm run test:update` to update snapshots, if needed (not needed)
  - [x] I've updated [docs](https://rjsf-team.github.io/react-jsonschema-form/docs) if needed (not needed)
  - [x] I've updated the [changelog](https://github.com/rjsf-team/react-jsonschema-form/blob/main/CHANGELOG.md) with a description of the PR
- [ ] **I'm adding a new feature**
  - [ ] I've updated the playground with an example use of the feature

### CODE REVIEW

File: packages/utils/src/findSchemaDefinition.ts [hunk 1]
```diff
@@ -18,37 +18,69 @@ export function splitKeyElementFromObject(key: string, object: GenericObjectType
...
  * @param $ref - The ref string for which the schema definition is desired
  * @param [rootSchema={}] - The root schema in which to search for the definition
+ * @param recurseList - List of $refs already resolved to prevent recursion
  * @returns - The sub-schema within the `rootSchema` which matches the `$ref` if it exists
- * @throws - Error indicating that no schema for that reference exists
+ * @throws - Error indicating that no schema for that reference could be resolved
  */
-export default function findSchemaDefinition<S extends StrictRJSFSchema = RJSFSchema>(
+export function findSchemaDefinitionRecursive<S extends StrictRJSFSchema = RJSFSchema>(
   $ref?: string,
```
Comment (heath-freenome) [inline packages/utils/src/findSchemaDefinition.ts:33]: By defaulting to an empty string here you can simplify the code below a little bit
```suggestion
  $ref?: string = '',
```

File: packages/utils/src/findSchemaDefinition.ts [hunk 2]
```diff
@@ -18,37 +18,69 @@ export function splitKeyElementFromObject(key: string, object: GenericObjectType
...
 
-/** Given the name of a `$ref` from within a schema, using the `rootSchema`, look up and return the sub-schema using the
- * path provided by that reference. If `#` is not the first character of the reference, or the path does not exist in
- * the schema, then throw an Error. Otherwise return the sub-schema. Also deals with nested `$ref`s in the sub-schema.
+/** Given the name of a `$ref` from within a schema, using the `rootSchema`, recursively look up and return the
+ * sub-schema using the path provided by that reference. If `#` is not the first character of the reference, the path
+ * does not exist in the schema, or the reference resolves circularly back to itself, then throw an Error.
+ * Otherwise return the sub-schema. Also deals with nested `$ref`s in the sub-schema.
  *
  * @param $ref - The ref string for which the schema definition is desired
```
Comment (heath-freenome) [inline packages/utils/src/findSchemaDefinition.ts:26]:
```suggestion
 * @param [$ref=''] - The ref string for which the schema definition is desired
```

**Comment by heath-freenome** [comment]
@MarttiR Can you fix the tests that are failing the pipeline?

**Comment by heath-freenome** [comment]
@MarttiR Also, can you rebase and resolve the conflicts?

**Comment by MarttiR** [comment]
@heath-freenome I've rebased the PR onto main.

The suggested changes modify the interface with respect to `$ref = undefined` values and break other tests. I don't really want to make any unnecessary changes here.

Having the explicit `decodedRef` variable is beneficial for readability – otherwise `ref` starts off as a string, but later becomes an object.

**Comment by heath-freenome** [comment]
@MarttiR I understand. Those were just suggestions. Let's see whether the update PR will build.

**Comment by heath-freenome** [comment]
> @heath-freenome I've rebased the PR onto main.
> 
> The suggested changes modify the interface with respect to `$ref = undefined` values and break other tests. I don't really want to make any unnecessary changes here.
> 
> Having the explicit `decodedRef` variable is beneficial for readability – otherwise `ref` starts off as a string, but later becomes an object.
> 
> The failing tests don't seem related to this PR; all `utils` tests are passing.

@MarttiR Can you deal deal with the conflicts so that we can get this merged assuming tests pass. Also, can you verify that all tests are passing in your branch on your local machine?

**Comment by MarttiR** [comment]
@heath-freenome Rebased the branch onto main. 

Tests were passing, but the test suite was indeed failing due to decreased coverage. I applied the same set of tests onto `findSchemaDefinitionRecursive` – could have gone fancy here, but just copy-pasting is more flexible in case the base function needs to diverge from the recursive one later.

Anyway, coverage is now 100%, and tests are passing.

>  Lerna (powered by Nx)   Successfully ran target test for 14 projects

**Review by heath-freenome — APPROVED** [review]

**Pull request merged**